File: dart_client/lib/src/services/questions_service.dart
import 'package:dio/dio.dart';
import '../dtos.dart';
import '../enums.dart';

class QuestionsService {
  final Dio _dio;
  QuestionsService(this._dio);

  Future<DTOResponse<QuestionListResponseDTO>> getQuestions({required int page, String? locale}) async {
    final response = await _dio.get('/questions', queryParameters: {'page': page, 'locale': locale, });
    return DTOResponse<QuestionListResponseDTO>.fromJson(response.data, (json) => QuestionListResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<GenericResponseDTO>> postQuestions({required CreateQuestionsRequestDTO body}) async {
    final response = await _dio.post('/questions', data: body.toJson());
    return DTOResponse<GenericResponseDTO>.fromJson(response.data, (json) => GenericResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<QuestionDataResponseDTO>> getQuestionsId({required int id, String? locale}) async {
    final response = await _dio.get('/questions/$id', queryParameters: {'locale': locale, });
    return DTOResponse<QuestionDataResponseDTO>.fromJson(response.data, (json) => QuestionDataResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<QuestionDataResponseDTO>> putQuestionsId({required UpdateQuestionRequestDTO body, required int id}) async {
    final response = await _dio.put('/questions/$id', data: body.toJson());
    return DTOResponse<QuestionDataResponseDTO>.fromJson(response.data, (json) => QuestionDataResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<GenericResponseDTO>> deleteQuestionsId({required int id}) async {
    final response = await _dio.delete('/questions/$id');
    return DTOResponse<GenericResponseDTO>.fromJson(response.data, (json) => GenericResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<QuestionListResponseDTO>> getBatch({required List<int> ids, String? locale}) async {
    final response = await _dio.get('/questions/batch', queryParameters: {'ids': ids, 'locale': locale, });
    return DTOResponse<QuestionListResponseDTO>.fromJson(response.data, (json) => QuestionListResponseDTO.fromJson(json as Map<String, dynamic>));
  }

}

File: js_client/contracts/services/questions_service.js
import { DTOResponse, DTOParams, EmptyParamsDTO, SearchQuestionsParamsDTO, GetQuestionParamsDTO, GetQuestionsBatchParamsDTO, UpdateQuestionParamsDTO, DeleteQuestionParamsDTO, UpdateCollectionParamsDTO, EmptyRequestDTO, BaseResponse, LoginResponseDTO, GenericResponseDTO, MessageResponseDTO, MustChangePasswordResponseDTO, UserStatusResponseDTO, ProfileDataResponseDTO, QuestionListResponseDTO, QuestionDataResponseDTO, CollectionListResponseDTO, IdDataResponseDTO, CollectionDetailResponseDTO, MarkListResponseDTO, FriendRequestResponseDTO, UserListResponseDTO, LoginRequestDTO, LoginDataDTO, SignupRequestDTO, MessageDataDTO, RecoverPasswordRequestDTO, VerifyEmailRequestDTO, MustChangePasswordDataDTO, ChangePasswordRequestDTO, UserStatusDataDTO, RegisterPushTokenRequestDTO, ProfileDataDTO, UpdateProfileRequestDTO, AnswerDataDTO, QuestionDataDTO, LocalizationDTO, CreateAnswerInputDTO, CreateQuestionInputDTO, CreateQuestionsRequestDTO, UpdateAnswerInputDTO, UpdateQuestionRequestDTO, CollectionDataDTO, CollectionDetailDataDTO, CreateCollectionRequestDTO, IdDataDTO, UpdateCollectionRequestDTO, MarkDataDTO, SendFriendRequestDTO, FriendRequestDataDTO, RespondFriendRequestRequestDTO, FriendRequestListResponseDTO, PublicUserProfileDTO, ErrorDetailsDTO } from '../dtos.js';
import * as Enums from '../enums.js';

export class QuestionsService {
  constructor(baseUrl, fetcher) {
    this.baseUrl = baseUrl;
    this.fetcher = fetcher;
  }

  async getQuestions({ page, locale,  } = {}) {
    const url = new URL(`${this.baseUrl}/questions`);
    if (page !== undefined && page !== null) url.searchParams.append('page', page);
    if (locale !== undefined && locale !== null) url.searchParams.append('locale', locale);
    const headers = { 'Content-Type': 'application/json' };
    const response = await this.fetcher(url.toString(), {
      method: 'GET',
      headers,
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => QuestionListResponseDTO.fromJson(data));
  }

  async postQuestions({ body, bearerToken } = {}) {
    const url = new URL(`${this.baseUrl}/questions`);
    const headers = { 'Content-Type': 'application/json' };
    if (bearerToken) headers['Authorization'] = `Bearer ${bearerToken}`;
    const response = await this.fetcher(url.toString(), {
      method: 'POST',
      headers,
      body: JSON.stringify(body.toJson())
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => GenericResponseDTO.fromJson(data));
  }

  async getQuestionsId({ id, locale,  } = {}) {
    const url = new URL(`${this.baseUrl}/questions/${encodeURIComponent(id)}`);
    if (locale !== undefined && locale !== null) url.searchParams.append('locale', locale);
    const headers = { 'Content-Type': 'application/json' };
    const response = await this.fetcher(url.toString(), {
      method: 'GET',
      headers,
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => QuestionDataResponseDTO.fromJson(data));
  }

  async putQuestionsId({ body, id, bearerToken } = {}) {
    const url = new URL(`${this.baseUrl}/questions/${encodeURIComponent(id)}`);
    const headers = { 'Content-Type': 'application/json' };
    if (bearerToken) headers['Authorization'] = `Bearer ${bearerToken}`;
    const response = await this.fetcher(url.toString(), {
      method: 'PUT',
      headers,
      body: JSON.stringify(body.toJson())
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => QuestionDataResponseDTO.fromJson(data));
  }

  async deleteQuestionsId({ id, bearerToken } = {}) {
    const url = new URL(`${this.baseUrl}/questions/${encodeURIComponent(id)}`);
    const headers = { 'Content-Type': 'application/json' };
    if (bearerToken) headers['Authorization'] = `Bearer ${bearerToken}`;
    const response = await this.fetcher(url.toString(), {
      method: 'DELETE',
      headers,
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => GenericResponseDTO.fromJson(data));
  }

  async getBatch({ ids, locale,  } = {}) {
    const url = new URL(`${this.baseUrl}/questions/batch`);
    if (ids !== undefined && ids !== null) url.searchParams.append('ids', ids);
    if (locale !== undefined && locale !== null) url.searchParams.append('locale', locale);
    const headers = { 'Content-Type': 'application/json' };
    const response = await this.fetcher(url.toString(), {
      method: 'GET',
      headers,
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => QuestionListResponseDTO.fromJson(data));
  }

}
